src/Microsoft.Diagnostics.Monitoring.WebApi/Stacks/SpeedScopeStacksFormatter.cs

```diff
@@ -0,0 +1,127 @@
+using System.IO;
+using System.Reflection;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.WebApi.Stacks
+{
+    internal sealed class SpeedscopeStacksFormatter : StacksFormatter
+    {
+        private const double FixedStart = 0.0;
+        private const double FixedEnd = 1.0;
+        private static readonly string Exporter = FormattableString.Invariant($"dotnetmonitor@{Assembly.GetExecutingAssembly().GetInformationalVersionString()}");
+        private const string Name = "speedscope.json";
+
+        public SpeedscopeStacksFormatter(Stream outputStream) : base(outputStream)
+        {
+        }
+
+        public override async Task FormatStack(CallStackResult stackResult, CancellationToken token)
+        {
+            // Speedscope contains a shared set of frames and each callstack references an index
+            // into the shared frame pool.
+            // We map each FunctionId to a corresponding index.
+            // Index 0 is reserved for Native frames
+            // Index 1 has no corresponding FunctionId and is reserved for Unknown.
+            var functionToSharedFrameMap = new Dictionary<ulong, int>()
+            {
+                {0, 0}
+            };
+
+            var speedscopeResult = new Models.SpeedscopeResult();
+
+            speedscopeResult.ActiveProfileIndex = 0;
+            speedscopeResult.Profiles = new List<Models.Profile>();
+            speedscopeResult.Exporter = Exporter;
+            speedscopeResult.Name = Name;
+            speedscopeResult.Shared = new Models.SharedFrames
+            {
+                Frames = new List<Models.SharedFrame>()
+                {
+                    new Models.SharedFrame{ Name = NativeFrame },
+                    new Models.SharedFrame{ Name = UnknownClass }
+                }
+            };
+
+            NameCache cache = stackResult.NameCache;
+
+            foreach (CallStack stack in stackResult.Stacks)
+            {
+                Models.Profile profile = new Profile()
+                {
+                    Events = new List<ProfileEvent>(),
+                    StartValue = FixedStart,
+                    EndValue = FixedEnd,
+                    Unit = UnitType.none,
+                    Name = string.Format(CultureInfo.CurrentCulture, Strings.CallstackThreadHeader, stack.ThreadId),
+                };
+
+                speedscopeResult.Profiles.Add(profile);
+
+                foreach (CallStackFrame frame in stack.Frames)
+                {
+                    if (frame.FunctionId == 0)
+                    {
+                        profile.Events.Add(new ProfileEvent { At = FixedStart, Frame = 0, Type = ProfileEventType.O });
+
+                    }
+                    else if (cache.FunctionData.TryGetValue(frame.FunctionId, out FunctionData functionData))
+                    {
+                        var builder = new StringBuilder();
```

Comment:
The construction of and appendation to the builder should only be done if the shared frame doesn't exist, otherwise this looks like a waste of compute. I think this will be fairly common occurrence (shared frames), especially for async call frames.